[ci][R-package] run R CRAN checks on Solaris by optional workflow

StrikerRUS opened the pull request:
Refer to #3740.

How it looks like: https://github.com/StrikerRUS/LightGBM/pull/5#issuecomment-773700926.

![image](https://user-images.githubusercontent.com/25141164/107103451-b9356280-682e-11eb-8cdd-b27d1385696d.png)

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,87 @@
+        )
+        , platform = platform
+        , env_vars = c(
+            "R_COMPILE_AND_INSTALL_PACKAGES" = "always"
+            , "_R_CHECK_SYSTEM_CLOCK_" = 0L
+            , "_R_CHECK_CRAN_INCOMING_REMOTE_" = 0L
+            , "_R_CHECK_PKG_SIZES_THRESHOLD_" = 60L
+            , "_R_CHECK_TOPLEVEL_FILES_" = 0L
```
These env variables should suppress some notes and warnings, but they don't...  ¯\_(ツ)_/¯

Possibly related: https://github.com/r-hub/rhub/issues/105 and https://github.com/r-hub/rhub/issues/174.

![image](https://user-images.githubusercontent.com/25141164/107103713-a5d6c700-682f-11eb-928b-01ed4995c897.png)

jameslamb replied:
😕 oy. Well, nothing we can do about it I guess

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,87 @@
+            , "_R_CHECK_PKG_SIZES_THRESHOLD_" = 60L
+            , "_R_CHECK_TOPLEVEL_FILES_" = 0L
+        )
+        , show_status = TRUE
```
Wait for `rhub::check()` finished.

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,87 @@
+        )
+        , show_status = TRUE
+    )
+    statuses <- res_object[[".__enclos_env__"]][["private"]][["status_"]]
```
Based on this great tutorial: https://jozef.io/r107-multiplatform-gitlabci-rhub/.

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,87 @@
+        note = iconv(note, "UTF-8", "ASCII", sub="")
+        # https://github.com/r-hub/rhub/issues/415
+        if (!(startsWith(note, "checking CRAN incoming feasibility")
+            || note == "checking compilation flags used ... NOTE\nCompilation used the following non-portable flag(s):\n  -march=pentiumpro")) {
```
Assert against full note body to ensure that **only** `march=pentiumpro` flag causes this note.

Refer to 

> I would be nervous about disabling the "checking compilation flags" one though, since that is system-dependent and could catch issue's in `R-package/configure`.
  https://github.com/StrikerRUS/LightGBM/pull/3#issuecomment-772893909

jameslamb replied:
makes sense to me!

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+    for (note in notes) {
+        note = iconv(note, "UTF-8", "ASCII", sub="")
```
`iconv()` to remove non-ASCII quotes around `march=pentiumpro`.

![image](https://user-images.githubusercontent.com/25141164/107103935-a7ed5580-6830-11eb-9ba4-f3eb2e329b51.png)


https://stackoverflow.com/a/38184834
https://stackoverflow.com/a/17292126

StrikerRUS commented on .github/workflows/r_solaris.yml:
```diff
@@ -0,0 +1,76 @@
+jobs:
+  test:
+    name: solaris-cran
+    timeout-minutes: 60
```
I run 20+ different checks on various platforms. All of them either finished within around 20 minutes or were aborted on RHub side (whole build took 44 minutes in that case):

```
3527#> * checking for unstated dependencies in ‘tests’ ... OK
3528#> * checking tests ...
3529#> Running ‘testthat.R’Build timed out (after 20 minutes). Marking the build as failed.
3530#> Build was aborted
```

jameslamb replied:
I'm surprised, but glad to hear it! I have had experiences before where I submit a job on a less-common platform like Solaris and it runs overnight (maybe 10 hours from submission to result)

StrikerRUS replied:
Maybe there were no timeouts on Rhub before?.. I don't know... You can check "Submitted:" and "Build time:" lines in all e-mails in http://www.yopmail.com/lightgbm_test_email to check response time.

![image](https://user-images.githubusercontent.com/25141164/107132429-23620c00-68f0-11eb-8fcf-52c7c6fb3927.png)

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,87 @@
+args <- commandArgs(
```
@jameslamb Please forgive me for this R code! 🙈 
Feel free to push any changes directly to the branch 😃 .

jameslamb replied:
it's very good, don't worry!

jameslamb commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,85 @@
+)
+rhub::validate_email(
+    email = intToUtf8(email - 42L)
+    , token = "6bc89147c8fc4824bce09f8454e4ab8e"
```
I haven't found any documentation on how often these tokens expire. But from my experience, I've found that about once every 6 weeks, I have to re-validate my email with R Hub. I don't know the rules for when and why validation is revoked, so I don't know for sure if hard-coding a token here is a problem.

I'm fine with leaving this here and seeing how it goes, but let's watch it carefully because base don my experience with R Hub, I expect we'll have to have a PR every two months or so to update this token.

StrikerRUS replied:
OK, let's leave this token here, I don't mind.
Let's hope that nobody will abuse RHub with our public e-mail address and publicly available token.

jameslamb replied:
if you want in a follow-up PR, you could make it a GitHub Secret (https://docs.github.com/en/actions/reference/encrypted-secrets#using-encrypted-secrets-in-a-workflow, mentioned in https://github.com/StrikerRUS/LightGBM/pull/3#issuecomment-772893909), set that as an environment variable, and read it in with `Sys.getenv()`. Then no one will see it in plain text.

Can be a later PR though

StrikerRUS replied:
Let's check how often we will have to update this token first. If not so often, then I'll ask you to add new token as a repo secret before preparing a new PR because I don't have an access for doing this.

But right after merging this PR there will be a small follow-up one with new e-mail/token (because these were for testing) and check that everything is working as expected.

jameslamb replied:
sounds good. Sorry, I missed sone of your comments and I created confusion because of it. I also do not have permission https://github.com/microsoft/LightGBM/pull/3913#discussion_r571502568

but I'm fine leaving this in for now

jameslamb commented on R-package/README.md:
````diff
@@ -354,6 +354,10 @@ rhub::check(
 )
 ```
 
+Alternatively, GitHub Actions can run code above for you. On a pull request, create a comment with this phrase:
+
+> /gha run r-solais
+
````
```suggestion
Alternatively, GitHub Actions can run code above for you. On a pull request, create a comment with this phrase:

> /gha run r-solais

**NOTE:** Please only do this only once you see that other R tests on a pull request are passing. R Hub is a free resource with limited capacity, and we want to be respectful community members.
```

Could you please add a note encouraging some restraint in the use of this job?

StrikerRUS replied:
Sure, will add this note.

jameslamb replied:
thank you

jameslamb reviewed (changes requested):
Thanks for doing this! I've left some initial comments.

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,85 @@
+email <- c(
+    150L, 147L, 145L, 146L, 158L, 145L, 140L, 151L, 137L, 158L, 143L, 157L, 158L, 137L, 143L, 151L,
+    139L, 147L, 150L, 106L, 163L, 153L, 154L, 151L, 139L, 147L, 150L, 88L, 141L, 153L, 151L
+)
```
Refer to https://github.com/microsoft/LightGBM/pull/3913#discussion_r571326426.

I'd prefer not to include plain e-mail address anywhere in the repo. It will prevent scrapping it by bots and in consequence full inbox of "your first BTC", "hello mister, I need help" and other very interesting e-mails. 😄 
In automatic comment script will post a **link to email** (http://www.yopmail.com/lightgbm_test_email) **not actually e-mail** itself (![image](https://user-images.githubusercontent.com/25141164/107131386-d75e9980-68e6-11eb-9f85-b30939c6ca7a.png)).

jameslamb replied:
ok guess I misread the GitHub Actions config

StrikerRUS replied:
I guess you remember some old version where this email indeed was hardcoded as plain text.

jameslamb replied:
I think some change in the GitHub UI confused me: https://github.com/microsoft/LightGBM/pull/3913#discussion_r571502568

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,85 @@
+log_file <- args[[2L]]
+dir.create(dirname(log_file), recursive = TRUE, showWarnings = FALSE)
+
+install.packages("rhub", dependencies = c("Depends", "Imports", "LinkingTo"), repos = "https://cran.r-project.org")
```
Ah OK, sure, will do it!

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,85 @@
+    , token = "6bc89147c8fc4824bce09f8454e4ab8e"
+)
+
+if (Sys.info()["sysname"] == "Windows") {
+    null_file <- "NUL"
+} else {
+    null_file <- "/dev/null"
+}
+sink(file = null_file)
```
Please refer to https://github.com/microsoft/LightGBM/pull/3913#discussion_r571327249.

jameslamb replied:
alright

StrikerRUS commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,85 @@
+        checks_succeeded <- FALSE
+    }
+    for (warning in warnings) {
+        warning <- iconv(x = warning, from = "UTF-8", to = "ASCII", sub = "")
```
Please refer to https://github.com/microsoft/LightGBM/pull/3913#discussion_r571328856.

jameslamb replied:
oh I see, did not see that comment. Sounds good

StrikerRUS commented on .github/workflows/r_solaris.yml:
```diff
@@ -0,0 +1,70 @@
+    timeout-minutes: 120
+    runs-on: ubuntu-latest
+    container: rocker/r-base
+    env:
+      SECRETS_WORKFLOW: ${{ secrets.WORKFLOW }}
```
It is needed to get optional workflows to work in general. For example, see this line: https://github.com/microsoft/LightGBM/blob/fc6b71e08ed89334a4fd859c03303619459a4168/.ci/append_comment.sh#L31.

jameslamb replied:
alright

StrikerRUS commented on .github/workflows/r_solaris.yml:
```diff
@@ -0,0 +1,70 @@
+              libxml2-dev \
+              libssl-dev
+          log_file="$GITHUB_WORKSPACE/rhub_logs.txt"
+          Rscript $GITHUB_WORKSPACE/.ci/run_rhub_solaris_checks.R "$(pwd)/lightgbm_${LGB_VERSION}.tar.gz" $log_file || exit -1
```
Not sure this will work, but I'll try.

StrikerRUS replied:
OK, you are right! This works.

jameslamb commented:
Looks fine, please `@` me whenever CI is passing and you'd like another review

jameslamb commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,87 @@
+)
+rhub::validate_email(
+    email = intToUtf8(email - 42L)
+    , token = '6bc89147c8fc4824bce09f8454e4ab8e'
```
ah I'm sorry, I didn't see this comment yesterday either!! It seems like I missed several of your comments. Maybe because I was looking on the "files" tab and these were on lines in the diff that had changed or something.

I would be happy to help, but I do not have permission on this repo to create secrets. I think only @guolinke can do that

jameslamb replied:
I noticed that all the comments I missed are on parts of the diff that say "outdated", so  guess it is a GitHub UI thing

![image](https://user-images.githubusercontent.com/7608904/107131879-30570f00-68a0-11eb-8c74-e246ba33bcbf.png)

StrikerRUS replied:
OK, no problem! Let's start with plain token in the source code, I think.

jameslamb commented on .ci/run_rhub_solaris_checks.R:
```diff
@@ -0,0 +1,87 @@
+email = c(
+    150L, 147L, 145L, 146L, 158L, 145L, 140L, 151L, 137L, 158L, 143L, 157L, 158L, 137L, 143L, 151L,
+    139L, 147L, 150L, 106L, 163L, 153L, 154L, 151L, 139L, 147L, 150L, 88L, 141L, 153L, 151L
+)
```
clever!

StrikerRUS commented:
@jameslamb I hope I addressed all your comments from previous review in 5e034a0d1f9473dd9cc2277596753e55de30a844 or gave my comment in the corresponding thread above.

jameslamb reviewed (approved):
Thanks for the hard work! I don't think many R projects have found a way to reliably and automatically test on Solaris. This will help our release confidence.

I think my concerns have been addressed. Let's merge this and see how it goes.
